_drafts/ember-orbit.md

```diff
@@ -0,0 +1,477 @@
+
+To make it appear, the first thing we need to do is to make the remote-update strategy non-blocking, like we did earlier for the remote-request:
+
+`app/data-strategies/remote-update.js`:
+
+```diff
+       target: 'remote',
+       action: 'push',
+
+-      blocking: true,
++      blocking: false,
+     });
+   }
+ };
+```
+
+Rerun the app and add a message, and it appears in the app.
+
+SCREENSHOT
+
+But this is **a lie**. This data isn't stored to the server-side API! This means that it won't be accessible on other devices, or to other users if the data is shared. And if the user clears their browser storage data, the data will be lost forever!
+
+Start the Rails API back up and reload the Ember app. The new message still appears locally. But let's check on the server as well. In a new terminal window, go to the API folder and run `bin/rails c`, then enter the command `Message.pluck(:text)`. The new message you added doesn't appear.
+
+SCREENSHOT
+
+This is kind of what we expect; it would be pretty advanced for an app to be able to keep track of updates made while offline and sync them up when it comes back online. But this is exactly what Orbit allows us to do!
+
+## IndexedDBBucket: Queueing Updates for When You Come Online
+
+We can use a feature of Orbit called a "bucket" to keep track of transformations that fail, so they can be retried later.
+
+First we create the bucket itself.
+
+`app/data-buckets/main.js`:
+
+```js
+import IndexedDBBucket from '@orbit/indexeddb-bucket';
+
+export default {
+  create() {
+    return new IndexedDBBucket({ namespace: 'my-app-settings' });
+  }
+};
+```
+
+Next, we need inject this bucket into our application so it's available to be used. We can do this in an Ember initializer:
+
+`app/initializers/orbit.js`:
+
+```js
+export function initialize(application) {
+  application.inject('data-source', 'bucket', 'data-bucket:main');
+  application.inject('data-key-map:main', 'bucket', 'data-bucket:main');
+}
+
+export default {
+  name: 'orbit',
+  initialize
+};
+```
+
+Now we can test to make sure in-flight requests are retried when you come online. Unfortunately the previous request you made is lost. But we can create a new one.
+
+Stop the Rails API. Reload the Ember app and create a new record. Start the Rails API, then reload the Ember app again. Amongst the console log entries you should see a `remote push` entry. Expand the details and you'll see that it's the record you added while offline. In the Rails console, run `Message.pluck(:text)` again, and you should see the record that was added while offline, synced up to the server with no effort from you!
+
+## Next Steps
+
+There are a few different places you can go from here.
+
+* Log truncation (WHAT DOES THIS ACCOMPLISH IN THIS CASE? I GET A NEW CONSOLE LOG ON EACH LOAD)
```

Comment:
?

Reply:
Log truncation strategies are explained [here](http://orbitjs.com/v0.15/guide/coordination.html#Log-truncation-strategies): 

> As changes are applied to sources, their transform logs grow in size. A log truncation strategy will keep the size in check by observing the sources associated with the strategy and truncating their logs when a common transform has been applied to them all.

The point is not to truncate console logs but rather the logs of transforms tracked at each source.

Reply:
OK, a couple of questions based on this:

1. Is there any relation between console logs and the `EventLoggingStrategy`? It seems as though [this commit](https://github.com/CodingItWrong/offline-ember/commit/a500e3a25fe338a47ec95d5b51d87ec69297bc01) that adds an `EventLoggingStrategy` to my sample app causes console logs to begin appearing.

2. Other than console logs, is there a way I can observe the difference between having an `EventLoggingStrategy` and not having it? Is there an error I'll get without the `EventLoggingStrategy`, or a feature I won't have available? Just trying to make the `EventLoggingStrategy` concrete.

Reply:
To answer both questions at once, the sole purpose of the `EventLoggingStrategy` is to log events to the console. This strategy should probably not be included with production builds.

Reply:
Wow, in hindsight, this is definitely a naming troll. Sorry for the confusion.

We should consider renaming `LogTruncationStrategy` (perhaps `TransformLogTruncationStrategy`?)  to be more clear.

But yes, any PRs that clarify this would be welcome. Thanks in advance.